# eBay Car Sales Data Cleaning

In this project, a dataset of used cars from eBay Kleinanzeigen, the classifieds section of the German eBay website, will be explored and cleaned. The dataset, originally scraped and uploaded to Kaggle by user orgesleka, includes a random sample of 50,000 data points from the full collection of listings.

Each entry in the dataset provides information about a car’s listing:
- dateCrawled - When the ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing.
- price - The listed selling price of the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The type of vehicle.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The type of transmission.
- powerPS - The power of the car in PS.
- model - The car model name.
- odometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

The goal of this project is to clean the dataset by handling any missing or inconsistent data, followed by performing an initial analysis to gain insights from the available information. By the end of the project, a cleaner, more usable dataset will allow for deeper exploration and analysis of trends in the used car market on eBay Kleinanzeigen.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'latin1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

The dataset consists of 20 columns, the majority of which are stored as string data types. A few columns contain missing values, though none have more than approximately 20% null values. Additionally, several columns have date information that is stored as strings.

To streamline the data for analysis, the first step will be to clean and standardize the column names, making the dataset easier to work with.

### Cleaning Columns

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

    gear_box  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_pictures  \
0     peugeot              nein  2016-03-26 00:00:00            0   
1         bmw              nein  2016-04-04 00:00:00            0   
2  volkswagen              nein  2016-03-26 00:00:00            0   
3       smart              nein  2016-03-12 00:00:00            0   
4        ford              nein  2016-04-01 00:00:00            0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

The column names in the dataset are written in camelCase, which is not in line with Python's preferred naming convention of snake_case. This means we cannot simply replace spaces with underscores. To improve readability and consistency, we'll convert all column names from camelCase to snake_case. Additionally, we will rename certain columns based on the data dictionary to make them more descriptive. For instance, yearOfRegistration will be renamed to registration_year, monthOfRegistration to registration_month, notRepairedDamage to unrepaired_damage, and dateCreated to ad_created. After renaming these specific columns, we will proceed to convert the remaining column names to snake_case and assign the modified names back to the autos.columns attribute.

### Data Exploration and Cleaning

The next step involves conducting basic data exploration to identify additional cleaning tasks. During this process, the focus will be on two main aspects. First, text columns where all or nearly all values are the same will be examined, as these columns often provide little to no useful information for analysis and can typically be dropped. Second, the exploration will look for instances of numeric data that are stored as text, which can be cleaned and converted to the appropriate numeric format for easier analysis and processing.

In [7]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year gear_box      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_pictures   postal_code            last_seen  
count                 50000      50000.0  50000.000000                50000  
unique                   76          NaN           NaN                39481  
top     2016-04-03 00:00:00          NaN           NaN  2016-04-07 06:17:27  
freq                   1946          NaN           NaN                    8  
mean                    NaN          0.0  50813.627300                  NaN  
std                     NaN          0.0  25779.747957                  NaN  
min                     NaN          0.0   1067.000000                  NaN  
25%                     NaN          0.0  30451.000000                  NaN  
50%                     NaN          0.0  49577.000000                  NaN  
75%                     NaN          0.0  71540.000000                  NaN  
max                     NaN          0.0  99998.000000                  NaN

In the dataset, both the seller and offer_type columns have only two unique values, with one value appearing 49,999 times out of 50,000 entries. This indicates that these columns offer little variability and may not provide meaningful insights for analysis. Additionally, there are several text columns where nearly all values are the same, including seller and offer_type, which may be candidates for removal.

The price and odometer columns, although numeric in nature, are currently stored as text. These columns will need to be cleaned by removing any non-numeric characters to properly convert them to numerical values. Another column, nr_pcitures, appears to have unusual data, requiring further investigation to understand its contents and determine how to handle it in the cleaning process.

In [8]:
autos["nr_pictures"].value_counts()

nr_pictures
0    50000
Name: count, dtype: int64

Since all the values in nr_pictures are 0, this column will be dropped as well as seller and offer_type.

In [9]:
autos = autos.drop(['nr_pictures', 'seller', 'offer_type'], axis=1)

In [10]:
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].str.replace(',', '')
autos['price'] = autos['price'].astype(int)

autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].astype(int)
autos.rename({'odometer':'odometer_km'}, axis = 1, inplace = True)

In [11]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year   gear_box  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

### Exploring the Odometer and Price Columns

In [12]:
print(autos['odometer_km'].value_counts())
print('\n')
#seeing how many unique values
print(autos['odometer_km'].unique().shape)
print('\n')
#viewing min/max/median/mean
print(autos['odometer_km'].describe())

odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: count, dtype: int64


(13,)


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


There are more high mileage than low mileage vehicles.

In [13]:
print(autos['price'].value_counts())
print('\n')
#seeing how many unique values
print(autos['price'].unique().shape)
print('\n')
#viewing min/max/median/mean
print(autos['price'].describe())

price
0        1421
500       781
1500      734
2500      643
1000      639
         ... 
16845       1
15470       1
18890       1
5198        1
1960        1
Name: count, Length: 2357, dtype: int64


(2357,)


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


In [14]:
autos['price'].value_counts().sort_index(ascending = True).head(20)

price
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: count, dtype: int64

There are 1,421 cars listed with a price of 0 in the dataset, which represents only about 2% of the total listings. Given the small proportion, these rows might be removed, as a price of $0 is likely incorrect or uninformative. Additionally, the dataset shows a maximum price of 100 million dollars, which seems unusually high for used cars. This warrants further investigation into the highest-priced listings to determine if they are outliers or errors that need to be addressed during the cleaning process.

In [15]:
autos['price'].value_counts().sort_index(ascending = False).head(20)

price
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: count, dtype: int64

The dataset also contains a few listings with extremely high prices, including 14 cars priced at or above $1 million. These numbers seem unrealistic for used cars, so it’s likely they are outliers or errors. It will be important to take a closer look at these listings to decide whether they should be removed or corrected during the data cleaning process.

Looking at the distribution of high prices in the dataset, there are some extreme values like `$99,999,999` and `$27,322,222`, which are clearly unrealistic for used cars. When reviewing the first 20 unique prices, we see that prices steadily rise up to about `$350,000`, but beyond that, they jump to much higher and less believable amounts. Because of this, we'll remove any listings with prices over `$350,000` to keep the analysis focused on more realistic values.

In [16]:
autos = autos[autos['price'].between(1,360000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

### Exploring the Date Columns

The dataset includes five columns that should represent date values. Some of these were generated by the crawler when the data was collected, while others come directly from the eBay website. It's important to properly format and handle these date columns, as they provide key details about when the listings were created, crawled, and last seen online:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

The date_crawled, last_seen, and ad_created columns are stored as strings, which means we'll need to convert them into a numerical format to analyze the data properly. The other two columns are already in numeric form, so we can use methods like Series.describe() to quickly understand their distribution without requiring any additional processing.

In [17]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [18]:
print(autos['date_crawled'].str[:10].value_counts(normalize = True, dropna = False).sort_index())

date_crawled
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: proportion, dtype: float64


In [19]:
print(autos['ad_created'].str[:10].value_counts(normalize = True, dropna = False).sort_index())

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: proportion, Length: 76, dtype: float64


In [20]:
print(autos['last_seen'].str[:10].value_counts(normalize = True, dropna = False).sort_index())

last_seen
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: proportion, dtype: float64


The crawler recorded the date it last observed each listing, which gives us a way to estimate when a car was removed from the site, likely due to being sold. However, we notice that the last three days in the dataset have a disproportionately high number of 'last seen' entries—about 6 to 10 times more than the days before. This suggests that it's unlikely there was a sudden surge in car sales. Instead, it's more probable that these values reflect the crawler's activity coming to an end, rather than actual sales of the vehicles.

The dataset shows a wide range of dates for when ads were created. While the majority of listings were created within 1-2 months of the crawler capturing the data, there are a few ads that are significantly older. In fact, the oldest ad dates back around 9 months. This variation suggests that while most cars were listed relatively recently, some had been on the site for a much longer period before being captured.

In [21]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year a car was first registered is an important indicator of its age, but there are some unusual values in this column. The minimum value is listed as 1000, which is clearly incorrect as cars did not exist then, and the maximum value is 9999, which is far in the future. These odd entries suggest there are errors in the data that will need to be addressed to ensure accurate analysis of the cars' ages.

### Incorrect Registration Year Data

From the exploration of the registration_year column, we noticed several odd values. The minimum year is listed as 1000, which is clearly incorrect since cars weren't invented yet, and the maximum year is 9999, far in the future. Since a car can't be registered after it was listed, any registration year after 2016 is inaccurate. While determining the earliest valid year is more challenging, it's likely within the first few decades of the 1900s. To clean this up, we'll count the number of listings with registration years outside the 1900–2016 range and decide whether we can remove those rows entirely, or if additional filtering logic is required.

In [22]:
# identify rows with registration years outside the range 1900-2016
invalid_years = ~autos["registration_year"].between(1900, 2016)

# count the number of invalid entries
invalid_count = invalid_years.sum()

# calculate the proportion of invalid entries
invalid_proportion = invalid_count / autos.shape[0]

invalid_proportion

np.float64(0.038793369710697)

Since a small portion of the data falls outside 1900-2016, that year range will be used for the rest of our exploring.

In [23]:
autos = autos[autos["registration_year"].between(1900, 2016)]

autos["registration_year"].value_counts(normalize = True).head(20)

registration_year
2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
Name: proportion, dtype: float64

It appears majority of cars were registered over the past 20 years.

### Exploring Price By Brand

The next step is to explore how prices vary across different car brands. By analyzing the price distribution for each brand, we can gain insights into which brands tend to have higher or lower average prices. This will help us understand market trends and identify any potential outliers or patterns in the dataset.

In [24]:
autos['brand'].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [25]:
autos['brand'].value_counts(normalize = True)

brand
volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.00

The dataset contains a wide range of car brands, with some brands being much more prevalent than others. Volkswagen is the most frequently listed brand, making up about 21% of the total listings, followed by BMW (11%) and Opel (10.8%). Other popular brands include Mercedes-Benz (9.6%), Audi (8.6%), and Ford (7%). Many other brands, such as Lada, Rover, and Saab, contribute less than 0.2% each to the overall listings.

To focus the analysis on more significant data, we will aggregate and analyze brands that account for more than 5% of the total listings. This will allow us to draw clearer insights from the major brands, while still retaining meaningful comparisons.

In [26]:
brand_perc = autos["brand"].value_counts(normalize=True)
common_brands = brand_perc[brand_perc > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object', name='brand')


In [27]:
avg_brand_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    avg_price = brand_only["price"].mean()
    avg_brand_prices[brand] = int(avg_price)
    
avg_brand_prices

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

After calculating the average prices for the most common car brands, we observe a distinct gap in price ranges. The average prices for Audi (`$9,336`), BMW (`$8,332`), and Mercedes-Benz (`$8,628`) are significantly higher compared to Ford (`$3,749`) and Opel (`$2,975`), which are more budget-friendly options. Volkswagen falls in the middle, with an average price of `$5,402`. This balance between affordability and quality may explain Volkswagen's popularity in the used car market.


### Aggregating By Mileage

or the top six car brands, we'll perform an aggregation to calculate the average mileage for each brand. This will help us explore whether there is any noticeable relationship between the average mileage of a brand's cars and their corresponding mean price. By comparing these two factors, we can gain insights into how mileage might influence the market value of different car brands.

In [28]:
bmp_series = pd.Series(avg_brand_prices)
pd.DataFrame(bmp_series, columns=["mean_price"])

mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749

In [29]:
avg_brand_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    avg_mileage = brand_only["odometer_km"].mean()
    avg_brand_mileage[brand] = int(avg_mileage)
    
avg_brand_mileage

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

In [30]:
mean_mileage = pd.Series(avg_brand_mileage).sort_values(ascending = False)
mean_prices = pd.Series(avg_brand_prices).sort_values(ascending = False)

In [31]:
brand_info = pd.DataFrame(mean_mileage,columns = ['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [32]:
brand_info["mean_price"] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

The range of car mileages across the top brands doesn't vary as significantly as the prices do. For the top brands, the mileages are all within a 10% range of each other. Interestingly, there is a slight trend where the more expensive vehicles, like Audi, BMW, and Mercedes-Benz, tend to have higher mileages, while the less expensive brands, such as Ford and Opel, have slightly lower mileages. This suggests that higher-priced cars may be driven more, but the difference in mileage is relatively small compared to the variation in prices. 

### Translating German Categorical Data

Several columns in the dataset contain German terms, which need to be translated into English for easier interpretation. Specifically, columns like vehicle_type, gear_box, and fuel_type included German values such as "manuell" (manual) and "benzin" (petrol). Using a mapping dictionary, these values will be translated into their English counterparts, making the dataset more accessible for further analysis.

In [33]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year   gear_box  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

In [34]:
# identify and translate German categorical data
german_to_english = {
    'vehicle_type': {
        'kleinwagen': 'small_car',
        'limousine': 'sedan',
        'cabrio': 'convertible',
        'coupe': 'coupe',
        'suv': 'suv',
        'andere': 'other',
        'bus': 'bus',
        'kombi': 'station_wagon'
    },
    'gear_box': {
        'manuell': 'manual',
        'automatik': 'automatic'
    },
    'fuel_type': {
        'benzin': 'petrol',
        'diesel': 'diesel',
        'elektro': 'electric',
        'hybrid': 'hybrid',
        'andere': 'other',
        'lpg': 'lpg',
        'cng': 'cng'
    },
    'unrepaired_damage': {
        'ja': 'yes',
        'nein': 'no'
        
    }
}

# Applying the translations
for column, translation_dict in german_to_english.items():
    if column in autos.columns:
        autos[column] = autos[column].map(translation_dict)


### Converting Dates to a Uniform Numeric Format

The dataset included multiple columns with dates stored as strings, such as date_crawled, last_seen, and ad_created. To standardize the format and make these dates easier to work with, the strings will be converted into an integer format (YYYYMMDD).

In [35]:
# convert dates to a uniform numeric format
date_columns = ['date_crawled', 'last_seen', 'ad_created']

for column in date_columns:
    if column in autos.columns:
        # extract only the date part before the space (ignore time)
        autos[column] = autos[column].str.split(' ').str[0]
        # convert date strings to integers in the format YYYYMMDD
        autos[column] = autos[column].str.replace('-', '').astype(int)

### Extracting Keywords from the name Column

The name column contains descriptive strings about each car listing. To extract more meaningful data from this column, the first word (typically the brand or model) will be pulled out as a new feature called name_keywords. 

In [36]:
# extract keywords from the 'name' column
autos['name_keywords'] = autos['name'].str.extract(r'(\w+)') 

autos.head()

date_crawled                                               name  price  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   5000   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500   
2      20160326                         Volkswagen_Golf_1.6_United   8990   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   4350   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   1350   

   ab_test   vehicle_type  registration_year   gear_box  power_ps   model  \
0  control            bus               2004     manual       158  andere   
1  control          sedan               1997  automatic       286     7er   
2     test          sedan               2009     manual       102    golf   
3  control      small_car               2007  automatic        71  fortwo   
4     test  station_wagon               2003     manual         0   focus   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   3       lpg     peugeot                no   
1       150000                   6    petrol         bmw                no   
2        70000                   7    petrol  volkswagen                no   
3        70000                   6    petrol       smart                no   
4       150000                   7    petrol        ford                no   

   ad_created  postal_code  last_seen  \
0    20160326        79588   20160406   
1    20160404        71034   20160406   
2    20160326        35394   20160406   
3    20160312        33729   20160315   
4    20160401        39218   20160401   

                                     name_keywords  
0                 Peugeot_807_160_NAVTECH_ON_BOARD  
1       BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  
2                                Volkswagen_Golf_1  
3                Smart_smart_fortwo_coupe_softouch  
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepflegt

### Find the Most Common Brand/Model Combinations

In this step, we will analyze the dataset to identify the most common brand and model combinations. By grouping the data based on both the car brand and model, we can determine which specific combinations appear most frequently in the listings. Sorting the results in descending order will help highlight the top 10 most popular brand/model combinations, giving us insights into which cars are most commonly listed on the platform.

In [37]:
common_brand_model = autos.groupby(['brand', 'model']).size().sort_values(ascending=False).head(10)
common_brand_model

brand          model   
volkswagen     golf        3707
bmw            3er         2615
volkswagen     polo        1609
opel           corsa       1592
volkswagen     passat      1349
opel           astra       1348
audi           a4          1231
mercedes_benz  c_klasse    1136
bmw            5er         1132
mercedes_benz  e_klasse     958
dtype: int64

The analysis of the most common brand and model combinations in the dataset reveals some interesting trends. Volkswagen dominates the listings, with the Volkswagen Golf being the most frequently listed car, appearing 3,707 times. Other popular models include the BMW 3 Series with 2,615 listings and the Volkswagen Polo with 1,609 listings. Opel also has a strong presence, with the Opel Corsa and Astra models showing 1,592 and 1,348 listings, respectively. Additionally, premium brands like Audi, BMW, and Mercedes-Benz are well-represented, with models such as the Audi A4, BMW 5 Series, and Mercedes-Benz C-Class and E-Class appearing frequently. This suggests a strong preference for Volkswagen, followed by a mix of both affordable and premium German car brands in the used car market.

### Analyze Price Patterns Based on Odometer Groups

Next, we will explore how car prices vary according to mileage by creating odometer groups. We will categorize the odometer_km values into defined bins (e.g., 0-50,000 km, 50,001-100,000 km, etc.) and then calculate the average price for each group. This analysis will allow us to observe any patterns or trends in price as mileage increases, providing insights into how a car's mileage impacts its market value.

In [38]:
# split the 'odometer_km' into groups and analyze average price patterns
odometer_bins = [0, 50000, 100000, 150000, 200000, 250000]
autos['odometer_group'] = pd.cut(autos['odometer_km'], odometer_bins)

# average price for each odometer group
odometer_price_pattern = autos.groupby('odometer_group', observed=True)['price'].mean()

odometer_price_pattern

odometer_group
(0, 50000]          14890.828856
(50000, 100000]      9595.419350
(100000, 150000]     4107.938641
Name: price, dtype: float64

The analysis of average car prices across different mileage groups reveals a clear trend. Cars with lower mileage, specifically those in the 0-50,000 km range, have the highest average price at `$14,890`. As mileage increases, prices drop significantly. Vehicles in the 50,000-100,000 km range have an average price of `$9,595`, while those with 100,000-150,000 km are priced much lower, with an average of `$4,108`. This indicates that mileage plays a key role in determining the resale value of cars, with lower-mileage vehicles commanding much higher prices compared to those with more wear and tear.

### Compare Prices Between Damaged and Non-Damaged Cars

Finally, we will analyze how much cheaper cars with unrepaired damage are compared to those without damage. By grouping the data based on the unrepaired_damage column and calculating the average price for each group, we will be able to quantify the price difference between damaged and non-damaged cars. This will provide a clear comparison of how unrepaired damage affects the resale value of vehicles in the dataset.

In [39]:
# Assume 'unrepaired_damage' column contains information about whether the car has unrepaired damage
damage_price_comparison = autos.groupby('unrepaired_damage')['price'].mean()
damage_price_comparison

unrepaired_damage
no     7164.033103
yes    2241.146035
Name: price, dtype: float64

The analysis comparing the prices of cars with and without unrepaired damage shows a significant difference in value. Cars with unrepaired damage have an average price of `$2,241`, while those without damage are priced much higher, with an average of `$7,164`. This suggests that unrepaired damage drastically reduces the resale value of a vehicle, highlighting the importance of a car's condition in determining its market price. Buyers are likely to pay a premium for cars that are in good condition and free from unrepaired issues.

### Conclusion

In this project, we explored and cleaned a dataset of used cars from eBay Kleinanzeigen. Throughout the process, we translated some categorical data from German to English, standardized date formats, and extracted relevant information from the name column to make the dataset more user-friendly for analysis.

Our analysis uncovered several key insights:
- Most Common Brand/Model Combinations: Volkswagen was the most frequently listed brand, with the Golf being the top model. Other popular models included the BMW 3 Series and Opel Corsa, showing that German car brands dominate the used car market.
- Mileage and Price Trends: There was a strong connection between a car’s mileage and its price. Cars with lower mileage (0-50,000 km) had significantly higher prices, while those with higher mileage saw their value drop sharply. This shows that buyers place a high value on cars with fewer kilometers driven.
- Impact of Unrepaired Damage on Price: Cars with unrepaired damage were much cheaper compared to those without damage. On average, cars with damage sold for about one-third of the price of undamaged cars, demonstrating how much a car’s condition affects its resale value.

Overall, this project provided a clear and structured approach to cleaning and analyzing the dataset, giving us insights into the factors that influence used car prices, such as brand, mileage, and condition.